src/lib/communication/DY.Lib.Communication.Core.fst

```diff
@@ -0,0 +1,278 @@
+  let* msg_id = send_msg msg_encrypted in
+  return (Some msg_id)
+
+
+val decrypt_message: bytes -> bytes -> option bytes
+let decrypt_message sk_receiver msg_encrypted =
+  pk_dec sk_receiver msg_encrypted
+
+val receive_confidential:
+  communication_keys_sess_ids ->
+  principal -> timestamp ->
+  traceful (option bytes)
+let receive_confidential comm_keys_ids receiver msg_id =
+  let*? sk_receiver = get_private_key receiver comm_keys_ids.private_keys (LongTermPkEncKey comm_layer_pkenc_tag) in
+  let*? msg_encrypted = recv_msg msg_id in
+  let*? payload = return (decrypt_message sk_receiver msg_encrypted) in
+  trigger_event receiver (CommConfReceiveMsg receiver payload);*
+  return (Some payload)
+
+
+(**** Authenticated Send and Receive Functions ****)
+
+val sign_message: principal -> principal -> bytes -> bytes -> bool -> bytes -> bytes -> bytes
+let sign_message sender receiver payload pk_receiver is_encrypted sk nonce =
+  let msg:communication_message = {sender; receiver; payload} in
+  if is_encrypted then (
+    let sig_input = Encrypted pk_receiver msg in
+    let sig_input_bytes = serialize signature_input sig_input in
+    let signature = sign sk nonce sig_input_bytes in
+    let signed_msg = {msg=sig_input_bytes; signature} in
+    serialize communication_message_sign signed_msg
+  ) else (
+    let sig_input = Plain msg in
+    let sig_input_bytes = serialize signature_input sig_input in
+    let signature = sign sk nonce sig_input_bytes in
+    let signed_msg = {msg=sig_input_bytes; signature} in
+    serialize communication_message_sign signed_msg
+  )
+
+val send_authenticated:
+  communication_keys_sess_ids ->
+  principal -> principal -> bytes ->
+  traceful (option timestamp)
+let send_authenticated comm_keys_ids sender receiver payload =
+  let*? sk_sender = get_private_key sender comm_keys_ids.private_keys (LongTermSigKey comm_layer_sign_tag) in
+  let* nonce = mk_rand SigNonce (long_term_key_label sender) 32 in
+  trigger_event sender (CommAuthSendMsg sender payload);*
+  let msg_signed = sign_message sender receiver payload empty false sk_sender nonce in
+  let* msg_id = send_msg msg_signed in
+  return (Some msg_id)
+
+#push-options "--ifuel 1 --fuel 0"
+val verify_message: principal -> bytes -> bool -> bytes -> bytes -> option communication_message
+let verify_message receiver sign_msg_bytes is_encrypted sk_receiver vk_sender =
+  let? msg_sign = parse communication_message_sign sign_msg_bytes in
+  let? sign_input = parse signature_input msg_sign.msg in
+  guard (verify vk_sender msg_sign.msg msg_sign.signature);?
+  if not is_encrypted then (
+    guard (Plain? sign_input);?
+    let Plain cm = sign_input in
+    guard (cm.receiver = receiver);?
+    Some cm
+  ) else (
+    guard (Encrypted? sign_input);?
+    let Encrypted pk_receiver cm = sign_input in
+    guard (cm.receiver = receiver);?
+    guard (pk_receiver = pk sk_receiver);?
+    Some cm
+  )
```

Comment:
```suggestion
  let (pk_receiver_opt, cm) = match sign_input with
    | Plain cm -> (None, cm)
    | Encrypted pk_receiver cm -> (Some pk_receiver, cm)
  in
  guard (cm.receiver = receiver);?
  guard (pk_receiver_opt = FStar.Option.mapTot pk sk_receiver_opt);
  Some cm
```
(Relies, of course, on changing the sk_receiver input to an option, at least as currently stated).